Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -225,150 +219,170 @@ - (UserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
+            }
+        });
     }
 }
 
 #pragma mark - Local Profile
 
 - (OWSAES128Key *)localProfileKey
 {
-    OWSAssert(self.localUserProfile.profileKey.keyData.length == kAES128_KeyByteLength);
+    @synchronized(self)
+    {
+        OWSAssert(self.localUserProfile.profileKey.keyData.length == kAES128_KeyByteLength);
 
-    return self.localUserProfile.profileKey;
+        return self.localUserProfile.profileKey;
+    }
 }
 
 - (BOOL)hasLocalProfile
 {
-    OWSAssert([NSThread isMainThread]);
-
     return (self.localProfileName.length > 0 || self.localProfileAvatarImage != nil);
 }
 
 - (nullable NSString *)localProfileName
 {
-    OWSAssert([NSThread isMainThread]);
-
-    return self.localUserProfile.profileName;
+    @synchronized(self)
+    {
+        return self.localUserProfile.profileName;
+    }
 }
 
 - (nullable UIImage *)localProfileAvatarImage
 {
-    OWSAssert([NSThread isMainThread]);
-
-    if (!self.localCachedAvatarImage) {
-        if (self.localUserProfile.avatarFileName) {
-            self.localCachedAvatarImage = [self loadProfileAvatarWithFilename:self.localUserProfile.avatarFileName];
+    @synchronized(self)
+    {
+        if (!self.localCachedAvatarImage) {
+            if (self.localUserProfile.avatarFileName) {
+                self.localCachedAvatarImage = [self loadProfileAvatarWithFilename:self.localUserProfile.avatarFileName];
+            }
         }
-    }
 
-    return self.localCachedAvatarImage;
+        return self.localCachedAvatarImage;
+    }
 }
 
 - (void)updateLocalProfileName:(nullable NSString *)profileName
                    avatarImage:(nullable UIImage *)avatarImage
-                       success:(void (^)())successBlock
+                       success:(void (^)())successBlockParameter
                        failure:(void (^)())failureBlockParameter
 {
-    OWSAssert([NSThread isMainThread]);
-    OWSAssert(successBlock);
+    OWSAssert(successBlockParameter);
     OWSAssert(failureBlockParameter);
 
-    // Ensure that the failure block is called on the main thread.
-    void (^failureBlock)() = ^{
-        dispatch_async(dispatch_get_main_queue(), ^{
-            failureBlockParameter();
-        });
-    };
-
-    // The final steps are to:
-    //
-    // * Try to update the service.
-    // * Update client state on success.
-    void (^tryToUpdateService)(NSString *_Nullable, NSString *_Nullable) = ^(
-        NSString *_Nullable avatarUrlPath, NSString *_Nullable avatarFileName) {
-        [self updateServiceWithProfileName:profileName
-            success:^{
-                // All reads and writes to user profiles should happen on the main thread.
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        @synchronized(self)
+        {
+            // Ensure that the success and failure blocks are called on the main thread.
+            void (^failureBlock)() = ^{
                 dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
DispatchMainThreadSafe() is an optimization - it avoids dispatching async if we're already on the main thread.  It comes at the cost of being slightly harder to read/understand.  I try to only use it where we need that optimization (e.g. to update a view synchronously if possible).